keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+- https://github.com/kubernetes/kubernetes/issues/87709 (logging of netpol actions, will help describing states we reach) 
+- https://github.com/projectcalico/felix/issues/2032 (non-deterministic time frames for policy applications - addressable through published performance tests which measure time to policy implementation for several pods in several namespaces.)
+- https://github.com/projectcalico/felix/issues/2008 (need to test postStart pods in networkpolicy upstream.)
+- https://github.com/vmware-tanzu/antrea/issues/381 (addressable by performance tests *or* node-targeted tests.)
+- https://github.com/kubernetes/kubernetes/issues/88375 (the test matrix for Egress is almost entirely empty, decreasing the verbosity of new tests will organically increase likelihood of new test submissions over time.)
+
+
+### Consequences of this problem
+ 
+The consequences of this problem is that
+ 
+- CNI providers cannot easily be compared for functionality.
+- CNI providers implementing network policies must carry a lot of downstream test functionality.
+- Testing a CNI provider for Kubernetes compatibility requires a lot of interpretation and time investment.
+- Extending NetworkPolicy tests is time consuming and error prone, without a structured review process and acceptance standard.
+- It is hard to debug tests, due to the performance characteristics - pods are deleted after each test, so we cannot reproduce the state of the cluster easily.
+
+## A high level outline of Pod Traffic Pathways
+
+Before diving into test details, we outline different types of pod traffic.  Each one of these types of traffic may pose different bugs to a CNI provider.  
+
+Intranode
+- pod -> pod
+- pod -> host 
+- host -> pod
+
+Internode
+- pod -> pod
+- pod -> host
+- host -> pod
+- `*`host -> host (out-of-scope for K8s Network Policies API)
+- host -> pod (host networking)
+
+Traffic Transiting Service DNAT
+- Nodeport -> service (DNAT) -> pod
+- pod -> service (DNAT) -> pod
+
+Intranamespace
+- pod -> pod
+- pod (host networking) -> pod
+- pod -> pod (host networking)
+
+Internamespace
+- pod -> pod
+- pod (host networking) -> pod
+- pod -> pod (host networking)
+
+## Detailed examples of the problem statement
+ 
+### Incompleteness
+
+A few concrete missing tests are obvious incompleteness examples, such as https://github.com/kubernetes/kubernetes/issues/87893 and https://github.com/kubernetes/kubernetes/issues/46625
+
+As mentioned in the pre-amble, there is sporadic validation of both positive and negative connectivity in all tests, and in many cases this validation is meaningful.  However, in none of the cases, it is complete.  That is, we do not have any tests which validate all obvious intra and inner namespace connectivity holes, both before and after application of policies.  
+
+Examples which visualize this ensue:
+
+For our first example, we will look at the incompleteness of one of the first tests
```

Comment:
Can you identify the specific test you're talking about?

And you need more detail here; it's not clear (a) what case the test claims to be testing, (b) what specific pieces of that it is actually testing, (c) what you think it should be testing but it's failing to

Reply:
sure, clarified it with the ginkgo sentence